### ISSUE TITLE
Discord Rich Presence layout could be improved

### ISSUE THREAD

**Issue by bramhaag** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#2978]
The current layout looks a bit weird in my opinion:
```
Google Play Music Desktop Player
Album {album} by {artists}
{song} ({x} out of {y})
xx:xx left
```

The `Album {album} by {artists}` line is in my opinion very confusing, the `{artists}` are the artists of the current song playing, not the "artist" of the album, i.e albums such as Super Eurobeat have songs from 10 or more artists. In my opinion, the layout could be improved by either making it fully customizable or by using something like [Discord's Spotify layout](https://owo.whats-th.is/34dd8b.png) (I am aware that this uses some restricted features)

**Comment by tankerkiller125** [comment]
I would like to see the format to be more like that of the Spotify Rich Presence

![Spotify Rich Presence](https://cdn-03.anonfile.com/99354bd0be/89d7af92-1517514377/Discord-Spotify-796x416.jpg)

**Comment by mrpotatofactory** [comment]
You should contact the Discord Team about "advanced features".
You probably could contact the Discord Team, you most likely would get the same treatment as Spotify did hopefully without royalties.

![features](https://img.synergyst.club/@potatofactory/chrome_2018-02-01_15-57-46.png)

**Comment by jostrander** [comment]
The Spotify integration appears to be a custom integration not using their standard RPC. I'm not sure that the advanced rich presence features would do any good in this case without contacting Discord for a similar layout.

**Comment by jostrander** [comment]
Also, @bramhaag the track artist is currently what's selected. We can only pull that data from the currently playing song area, so if that's what Google displays there our hands are tied.

https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/master/src/main/features/core/discordRichPresence.js#L51

**Comment by Alipoodle** [comment]
#2970 I've already mentioned the problem with it being cluttered, and even refereed back to the Discord Wiki... And still feel that some options should be made to make it seem less cluttered, especially with the amount of _useless_ information that is thrown at the user.

Stuff like ``Listening to Google Play Music Desktop Player`` and Album Art Is not possible due to using the Rich Presences API which only meant to be supported for Games (therefore playing) and images shouldn't be made on the fly / grabbed from another source..

Spotify is custom as per what jostrander said, and won't be coming to Google Play Music without some partnership. I doubt that Discord will create some partnership with just this application dude to it not getting the information from _an official source_.

Advanced Rich Presences features won't work either, as these are again required for games.

**Comment by aidant** [comment]
I agree it needs some work, If we could get listening integration with discord that would be awesome. I was thinking of removing the Album as it can get awfully cluttered and killing the presence after plenty of idling. If you have any suggestions id be happy to look into it.

**Comment by mrpotatofactory** [comment]
Instead of making the album art the generic logo, would it be possible to use the genre of music? Or does Google not provide such metadata?

**Comment by mrpotatofactory** [comment]
Then only use the default logo when the queue is idle

**Comment by TeRRa4** [comment]
I changed it to this, but I can't build it on any machine I have access to:
```javascript
    const presence = {
      state: track.artist,
      details: track.title,
      startTimestamp: start,
      endTimestamp: end,
      instance: false,
      largeImageKey: 'playing',
      largeImageText: 'Google Play Music',
    };
```

So in theory it should come out:
```
Google Play Music Desktop Player
Song name
Artist name
Duration remaining
```

**Comment by Techman-** [comment]
I wanted to bump this issue because I too would like to see the integration improved slightly. It would be awesome if the icon would change to match the album image, with a little beacon in the corner with the current logo that is shown, [kind of like the package for the Atom editor](https://github.com/HelloWorld017/atom-discord).
![image](https://user-images.githubusercontent.com/5396240/42725580-d8b792d6-8753-11e8-8245-c5102b024c7e.png)

**Comment by jostrander** [comment]
Near impossible @Techman- .  You have to upload the "album images" before hand and pick which one to show in the integration by name. the Atom editor can get away with it, because there is a rather finite list of programming languages etc to create artwork for and maintain.

**Comment by Techman-** [comment]
That's unfortunate. Is this worth bringing up in the Discord developer channels?

**Comment by jostrander** [comment]
I doubt they are going to allow random image art via URL for rich presence as it's too insecure.

**Comment by Alipoodle** [comment]
```
Album Art Is not possible due to using the Rich Presences API which only meant to be supported for Games (therefore playing) and images shouldn't be made on the fly / grabbed from another source..
```
Already mention this.
To get the album art we would need to get some form of partnership thing but as I've also mentioned:
```
Spotify is custom... , and won't be coming to Google Play Music without some partnership. I doubt that Discord will create some partnership with just this application dude to it not getting the information from an official source.
```

**Comment by Techman-** [comment]
What the developers could do is make a little "branch" of this API specifically for media players. Music apps can be individually cleared by a human if security is a concern.

Just floating ideas.

**Comment by Alipoodle** [comment]
> and won't be coming to Google Play Music without some partnership. I doubt that Discord will create some partnership with just this application dude to it not getting the information from an official source.

**Comment by Alipoodle** [comment]
By the sounds of this... it's not relevant to GPMDP.

Discord related stuff [Rich Presences](https://discordapp.com/rich-presence)
GPMDP JSON API [Docs](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/master/docs/PlaybackAPI.md)
GPMDP Web Socket API [Docs](https://github.com/MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-/blob/master/docs/PlaybackAPI_WebSocket.md)

**Comment by Techman-** [comment]
It's a general solution to a very local issue :P

**Comment by tf2manu994** [comment]
Also, it says playing but it should say 'listening to', like spotify.

**Comment by Alipoodle** [comment]
> Stuff like `Listening to Google Play Music Desktop Player`... Is not possible due to using the Rich Presences API which only meant to be supported for Games therefore playing instead of listening

**Comment by tf2manu994** [comment]
I know in discord.js you can set the type to 2 and then it says that

**Comment by Alipoodle** [comment]
While discord-rpc is an extension of D.js, they don't work in the same way at all!
Discord RPC doesn't allow for providing of type unlike D.js which will provide it through the API.

For a person who has worked on other Discord RPC related projects and D.js projects, I'm aware of the limitations within discord RPC.

*No Custom Album art images,  No `Listening to Google Play Music`, No Listening party, No in chat stuff*
Without some form of partnership with Discord - which is unlikely due to being a third party application.

Can we make this more relevant to the original issue and talk about it's layout!
